### PR TITLE
dice loss

### PR DESCRIPTION
**Pull request by xiexinch**

### CODE REVIEW

**Comment by xvjiarui** [comment]
Hi @xiexinch 
Please follow https://github.com/open-mmlab/mmsegmentation/blob/master/.github/CONTRIBUTING.md to fix the linting error.

File: mmseg/models/losses/dice_loss.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    num = torch.sum(torch.mul(pred, target) * valid_mask, dim=1) * 2 + smooth
+    den = torch.sum((pred.pow(exponent) + target.pow(exponent)) * valid_mask, dim=1) + smooth
+
+    return 1 - num / den
+
+@LOSSES.register_module()
+class DiceLoss(nn.Module):
+    """DiceLoss.
+
+    """
```
Comment (xvjiarui) [inline mmseg/models/losses/dice_loss.py:48]: We may add some docstring here.

File: mmseg/models/losses/dice_loss.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+                     smooth=1, 
+                     exponent=2,
+                     **kwards):
+    assert pred.shape[0] == target.shape[0]
+    pred = pred.contiguous().view(pred.shape[0], -1)
+    target = target.contiguous().view(target.shape[0], -1)
+    valid_mask = valid_mask.contiguous().view(valid_mask.shape[0], -1)
+
+    num = torch.sum(torch.mul(pred, target) * valid_mask, dim=1) * 2 + smooth
+    den = torch.sum((pred.pow(exponent) + target.pow(exponent)) * valid_mask, dim=1) + smooth
```
Comment (xvjiarui) [inline mmseg/models/losses/dice_loss.py:40]: We may directly use `denominator`.

**Review by xvjiarui — APPROVED** [review]

File: mmseg/models/losses/dice_loss.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
+import torch
```
Comment (xvjiarui) [inline mmseg/models/losses/dice_loss.py:1]: We may add sth like ``modified from xxx``.

File: mmseg/models/losses/dice_loss.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+            Default: 1
+        exponent (float): An float number to calculate denominator
+            value: \\sum{x^exponent} + \\sum{y^exponent}. Default: 2.
+        reduction (str, optional): The method used to reduce the loss. Options
+            are "none", "mean" and "sum". This parameter only works when
+            per_image is True. Default: 'mean'.
+        class_weight (list[float], optional): The weight for each class.
+            Default: None.
+        loss_weight (float, optional): Weight of the loss. Default to 1.0.
+        ignore_index (int | None): The label index to be ignored. Default: -1.
```
Comment (xvjiarui) [inline mmseg/models/losses/dice_loss.py:67]: Should be 255.

**Comment by xvjiarui** [comment]
FCN dilate model shouldn't be in this PR

File: mmseg/models/losses/dice_loss.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    def forward(self, pred, target, avg_factor=None, reduction_override=None):
+        assert reduction_override in (None, 'none', 'mean', 'sum')
+        reduction = (
+            reduction_override if reduction_override else self.reduction)
+        if self.class_weight is not None:
+            class_weight = pred.new_tensor(self.class_weight)
+        else:
+            class_weight = None
+
+        pred = F.softmax(pred, dim=1)
```
Comment (ZwwWayne) [inline mmseg/models/losses/dice_loss.py:102]: In instance segmentation, dice loss may use sigmoid for activation. Suggest supporting both cases.

File: mmseg/models/losses/dice_loss.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+              target,
+              valid_mask,
+              smooth=1,
+              exponent=2,
+              class_weight=None,
+              ignore_index=-1):
+    assert pred.shape[0] == target.shape[0]
+    total_loss = 0
+    num_classes = pred.shape[1]
+    for i in range(num_classes):
```
Comment (ZwwWayne) [inline mmseg/models/losses/dice_loss.py:22]: Use for loop might be inefficient? Some implementation support to process multi-class in a batched manner.

File: mmseg/models/losses/dice_loss.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+            if class_weight is not None:
+                dice_loss *= class_weight[i]
+            total_loss += dice_loss
+    return total_loss / num_classes
+
+
+@weighted_loss
+def binary_dice_loss(pred, target, valid_mask, smooth=1, exponent=2, **kwards):
+    assert pred.shape[0] == target.shape[0]
+    pred = pred.contiguous().view(pred.shape[0], -1)
```
Comment (ZwwWayne) [inline mmseg/models/losses/dice_loss.py:39]: `.contiguous().view()` can be replaced by `reshape`?

**Review by ZwwWayne — APPROVED** [review]

File: mmseg/models/losses/dice_loss.py (hunk 6 above)
Comment (ZwwWayne) [reply mmseg/models/losses/dice_loss.py:22]: https://github.com/LIVIAETS/boundary-loss/blob/master/losses.py
